### PR TITLE
Making batch client side changes to get nextstreamcut as per review comments

### PR DESCRIPTION
**Pull request by AnkurAnand11**
**Change log description**  
While calling the getNextStreamCut api, if we get the exception like NoSuchSegmentsException that does not really make sense to send back this to client as they are providing a streamCut so the segment must have existed at some point. So if it no longer exists, perhaps a SegmentTruncatedException would be more appropriate. Instead of creating a new wireCommand IndexSegmentSearchFailed, it would be better to handle the exception occurring while locating the offset for the segment.

**Purpose of the change**  
To Address the review comment given on https://github.com/pravega/pravega/commit/14af0d3cc7a6caf1797e6b1d66e950af147ab061

**What the code does**  
- In case of NoSuchSegmentException, response is being sent as SegmentTruncatedException.
- Removed the new wirecommand "IndexSegmentSearchFailed".
- In case of exception ( IllegalStateException/SearchFailedException) being thrown while locating the offset, catching the exception and sending the tail offset for the segment in such case.

**How to verify it**  
N/A

### CODE REVIEW

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java [hunk 1]
```diff
@@ -51,10 +60,19 @@ public SearchFailedException(String message) {
...
      */
-    public static long locateOffsetForSegment(StreamSegmentStore store, String segment, long targetOffset, boolean greater) throws SearchFailedException {
-        return applySearch(store, segment, targetOffset, greater).getValue();
+    public static long locateOffsetForSegment(StreamSegmentStore store, String segment, long targetOffset, boolean greater) throws StreamSegmentNotExistsException {
+        try {
+            long offset = applySearch(store, segment, targetOffset, greater).getValue();
+            return offset;
+        } catch (SegmentTruncatedException ex) {
+            throw new StreamSegmentNotExistsException(ex.getMessage());
+        } catch (IllegalStateException | SearchFailedException ex) {
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java:71]: Let's not catch IllegalState.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java [hunk 2]
```diff
@@ -51,10 +60,19 @@ public SearchFailedException(String message) {
...
+            log.warn("Sending the tail offset for the segment as it failed to locate offset for segment: {}", segment, ex);
+            SegmentProperties segmentProperties = store.getStreamSegmentInfo(segment, TIMEOUT).join();
+            return segmentProperties.getLength();
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java:74]: This is not the correct value. We should propagate an error.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java [hunk 3]
```diff
@@ -94,7 +112,9 @@ private static Map.Entry<Long, Long> applySearch(StreamSegmentStore store, Strin
                     return entry.getOffset();
                 case Future:
                 case EndOfStreamSegment:
+                    throw new IllegalStateException(String.format("Unexpected size of index segment of type: %s was encountered for segment %s.", firstElement.getType(), segment));
                 case Truncated:
+                    throw new SegmentTruncatedException(String.format("Segment %s no longer exists.", segment));
                 default:
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java:118]: This is an enum. I believe if the switch is exhaustive we can delete the default case.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java (hunk 1 above)
Comment (AnkurAnand11) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java]: Removed IllegalState catch block.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java (hunk 2 above)
Comment (AnkurAnand11) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java]: Removed the catch block.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java (hunk 3 above)
Comment (AnkurAnand11) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java]: Removed the default case and also SearchFailedException.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java [hunk 4]
```diff
@@ -51,13 +53,17 @@ public SearchFailedException(String message) {
...
+     * @throws StreamSegmentNotExistsException If the segment is truncated or it does not exist.
      */
-    public static long locateOffsetForSegment(StreamSegmentStore store, String segment, long targetOffset, boolean greater) throws SearchFailedException {
-        return applySearch(store, segment, targetOffset, greater).getValue();
+    public static long locateOffsetForSegment(StreamSegmentStore store, String segment, long targetOffset, boolean greater) throws StreamSegmentNotExistsException {
+        try {
+            long offset = applySearch(store, segment, targetOffset, greater).getValue();
+            return offset;
+        } catch (SegmentTruncatedException ex) {
+            throw new StreamSegmentNotExistsException(ex.getMessage());
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java:63]: These are different errors. Truncated means the segment exists but the data in it has been deleted, `StreamSegmentNotExistsException` means the segment itself does not exist. It doesn't make sense to covert between them like this.

Comment (AnkurAnand11) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java]: Reverted.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java (hunk 3 above)
Comment (AnkurAnand11) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java]: Done.

**Review by tkaitchuck — APPROVED** [review]

**Pull request merged**